Add Keylog file with env variable SSLKEYLOGFILE (for openssl)

gcolin opened the pull request:
Hello, this PR adds the generation of keylog file for debugging TLS with wireshark. This is working fine with openssl.

The env variable SSLKEYLOGFILE is used in others apps such as Curl, Firefox and Google Chrome.

michalvasko commented on src/keylog.c:
```diff
@@ -0,0 +1,72 @@
+#include <stdlib.h>
+#include "keylog.h"
+
+#define KEYLOG_LABEL_MAXLEN (sizeof("CLIENT_HANDSHAKE_TRAFFIC_SECRET") - 1)
+
+#define CLIENT_RANDOM_SIZE  32
+
+/*
+ * The master secret in TLS 1.2 and before is always 48 bytes. In TLS 1.3, the
+ * secret size depends on the cipher suite's hash function which is 32 bytes
+ * for SHA-256 and 48 bytes for SHA-384.
+ */
+#define SECRET_MAXLEN       48
```
What are these defines for? They do not seem to be used anywhere.

michalvasko commented on src/session_openssl.c:
```diff
@@ -43,6 +44,36 @@
+void
+tls_keylog_write_line(const SSL *ssl, const char *line)
+{
+    (void)ssl;
```
We use `UNUSED()` for the parameter instead.

gcolin replied:
ok done

michalvasko reviewed (changes requested):
Thanks for the PR, but to follow our code organization, a few changes will be required. 

Starting with some minor issues, we generally use the `nc_` prefix for all *libnetconf2* functions (possible exception only for static ones) to set a "namespace" differentiating between all the used libraries. But, since this is OpenSSL-only, I think it can all be put into the `session_openssl.c` file without introducing additional ones. Ideally, there should be a TLS-specific `init` function (for example), which will not do anything for the mbedTLS but initializes this keylogger for OpenSSL to follow our TLS library wrappers.

gcolin commented:
Hello Michal, I updated the PR. the code is only in session_openssl.c

I added 2 global variable in file session_openssl.c: active_sessions and keylog_file_fp. maybe there is a better place to store these global variables.

Roytak commented on src/session_openssl.c:
```diff
@@ -43,6 +43,73 @@
+FILE *keylog_file_fp = NULL;
+uint32_t active_sessions = 0;
+
+FILE*
```
Use static and add the missing space, so it's `static FILE *`.

Roytak commented on src/session_openssl.c:
```diff
@@ -43,6 +43,73 @@
+FILE*
+nc_tls_keylog_open(void)
+{
+    char *keylog_file_name = getenv(SSLKEYLOGFILE_ENV);
```
Separate the variable definition and the assignment.

Roytak commented on src/session_openssl.c:
```diff
@@ -43,6 +43,73 @@
+    return keylog_file_fp;
+}
+
+void
```
Use static.

Roytak commented on src/session_openssl.c:
```diff
@@ -43,6 +43,73 @@
+void
+nc_tls_keylog_close(void)
+{
+    char *keylog_file_name = getenv(SSLKEYLOGFILE_ENV);
```
Separate the variable definition and the assignment.

Roytak commented on src/session_openssl.c:
```diff
@@ -43,6 +43,73 @@
+    }
+}
+
+void
```
Use static.

Roytak commented on src/session_openssl.c:
```diff
@@ -43,6 +43,73 @@
+nc_tls_keylog_close(void)
+{
+    char *keylog_file_name = getenv(SSLKEYLOGFILE_ENV);
+    if (keylog_file_name) {
```
Could also check the value of the `keylog_file_fp` variable, so that this variable may be redundant if I'm not mistaken.

Roytak reviewed (commented):
Hi, thank you for the contribution. Can you change the target branch to `devel`? Also would be great if you could run `make format` once you fix the small details I commented on. I think it will be good to merge then, if all tests pass.

gcolin commented:
There is a mutex to fix a concurrency issue in test_tls: the file was open twice.

I cannot change the "from" nokia:master. Should I recreate a new PR?

Roytak commented:
Is there a specific reason as to why the filename is passed as an env variable or that's just how curl handles it? I am thinking that setting it via an API may be better (and possibly doing this in netopeer2). Anyways, it will most likely be merged soon. I'll try to find out if this can be done with mbedtls as well, and possibly edit it accordingly.

gcolin commented:
> Is there a specific reason as to why the filename is passed as an env variable or that's just how curl handles it? I am thinking that setting it via an API may be better (and possibly doing this in netopeer2). Anyways, it will most likely be merged soon. I'll try to find out if this can be done with mbedtls as well, and possibly edit it accordingly.

Hi, yes it is simpler:
. there is only one binary to update
. curl uses it. In our software, we use both curl and netopeer2
. other popular tools use this way

A colleague also suggested to have a paramater from netopeer2. For that, netopeer2 needs also an update.

That is cool to have the same in mbedtls.

Roytak commented:
Hello, I decided to implement it since it required a bit more changes and I liked this enhancement. Sorry for not merging yours. Nonetheless, feel free to try it out and give us some feedback if need be. Thank you.
